### PR TITLE
Add Google Cloud Storage repository plugin

### LINKED ISSUE #12880: Add Google Cloud Storage Repository
_From @dadoonet on January 28, 2014 21:18_

We want to have Snapshot and Restore available for Google Storage Repository.

_Copied from original issue: elastic/elasticsearch-cloud-gce#11_

### PR DESCRIPTION
**Pull request by tlrx**
This pull request adds a new type of repository `gcs` to allow to snapshot/restore indices on Google Cloud Storage service:

```
curl -XPUT 'localhost:9200/_snapshot/my_gcs_repo' -d '                       
{                 
  "type": "gcs", 
  "settings":  {
      "project_id": "my-project-id", 
      "bucket":"my-bucket", 
      "credentials": "/path/to/service-account-file.json"
   }
}
```

For now it only supports authentication using [Service Account](https://cloud.google.com/storage/docs/authentication#service_accounts) files.

Many things still need to be done but I'd be glad to benefit from a first round of review.
- [x] add integration tests
- [x] add license files
- [x] add documentation
- [ ] add HTTP proxy support
- [ ] add HTTP proxy support with custom java key stores

Closes #12880

### CODE REVIEW

**Comment by tlrx** [comment]
Note: it needs #13574 to work

File: plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+public class GoogleCloudStorageBlobContainer extends AbstractBlobContainer {
+
+    protected static final ESLogger logger = Loggers.getLogger(GoogleCloudStorageBlobContainer.class);
+
+    private final GoogleCloudStorageBlobStore blobStore;
+    private final String keyPath;
+
+    @Inject
+    protected GoogleCloudStorageBlobContainer(BlobPath path, GoogleCloudStorageBlobStore gcsBlobStore) {
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java]: I don't think we ever inject container, so, it shouldn't be annotated. We can also make the constructor package private, since we don't really inherit from GoogleCloudStorageBlobContainer. Lastly, why don't we just pass bucket and client since this is the only part of the store that we need here and it will simplify code below a bit.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/plugin/repository/gcs/GoogleCloudStorageModule.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+public class GoogleCloudStorageModule extends AbstractModule {
+
+    @Override
+    protected void configure() {
+        bind(GoogleCloudStorageService.class).to(gcsServiceClass()).asEagerSingleton();
+    }
+
+    Class<? extends GoogleCloudStorageService> gcsServiceClass() {
+        return InternalGoogleCloudStorageService.class;
+    }
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/plugin/repository/gcs/GoogleCloudStorageModule.java]: Is this a hook for future testing?

Comment (tlrx) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/plugin/repository/gcs/GoogleCloudStorageModule.java]: Yes

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+    @Override
+    protected void doStop() {
+        logger.trace("Google Cloud Storage service stopped");
+    }
+
+    @Override
+    protected void doClose() {
+        logger.trace("Google Cloud Storage service closed");
+    }
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java]: Since we are not doing anything in lifecycle methods, why make it into AbstractLifecycleComponent instead of just AbstractComponent in the first place?

**Comment by imotov** [comment]
Looks like a good start. Left a couple of comments.

**Comment by tlrx** [comment]
@imotov Thanks for your review. I updated the code following your comments and I also rebased the pull request so that it now works with gradle and the security manager.

Once #14050 is in I should be able to add tests.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/plugin/repository/gcs/GoogleCloudStorageModule.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+import org.elasticsearch.common.inject.AbstractModule;
+import org.elasticsearch.repositories.gcs.GoogleCloudStorageService;
+import org.elasticsearch.repositories.gcs.InternalGoogleCloudStorageService;
+
+public class GoogleCloudStorageModule extends AbstractModule {
+
+    @Override
+    protected void configure() {
+        bind(GoogleCloudStorageService.class).to(InternalGoogleCloudStorageService.class).asEagerSingleton();
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/plugin/repository/gcs/GoogleCloudStorageModule.java]: Since we bind to concrete implementation here, maybe it makes sense to get rid of GoogleCloudStorageService interface all together and rename InternalGoogleCloudStorageService to GoogleCloudStorageService. It feels like these interfaces don't buy us anything except added complexity. Replacing them with mock implementation would make little sense here since we already test all upper levels with mock implementation based on fs repo.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+            logger.debug("checking if bucket [{}] exists", bucketName);
+            try {
+                Bucket bucket = client.buckets().get(bucketName).execute();
+                if (bucket != null) {
+                    return Strings.hasText(bucket.getId());
+                }
+            } catch (GoogleJsonResponseException e) {
+                if (e.getStatusCode() == HTTP_FORBIDDEN) {
+                    logger.warn("access to bucket [{}] is forbidden", bucketName);
+                    return false;
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java]: I am not sure that false is correct answer here since if the bucket doesn't exist we will try to create it in the next step. Would it make more sense to bail out with an exception right here.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        client.deleteBlobs(bucket, buildKeys(blobNames));
+    }
+
+    @Override
+    public void move(String sourceBlobName, String targetBlobName) throws IOException {
+        client.moveBlob(bucket, buildKey(sourceBlobName), bucket, buildKey(targetBlobName));
+    }
+
+    protected String buildKey(String blobName) {
+        return keyPath + (blobName == null ? "" : blobName);
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java]: I don't think we should try to work `null` here. If we got `null` here it might indicate a more serious problem and I don't think hiding it by replacing `null` with `""` is the right solutions here. I think we should just bailout here with NPE.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+                        public void onSuccess(Void aVoid, HttpHeaders responseHeaders) throws IOException {
+                            logger.trace("successfully deleted blob [{}] in bucket [{}]", blobName, bucketName);
+                        }
+                    });
+
+                    count += 1;
+                } catch (IOException e) {
+                    logger.error("Unable to queue delete request for blob [{}] in bucket [{}]", e, blobName, bucketName);
+                } finally {
+                    it.remove();
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java]: That seems brittle (it assumes that collection we got is mutable) and not very useful (we don't rely on the fact that we deleted processed blobs from the collection).

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java [hunk 8]
```diff
@@ -0,0 +1,330 @@
...
+                }
+
+                if ((count >= MAX_BATCHING_REQUESTS)  || (it.hasNext() == false)) {
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java]: Should we reset count to 0 somewhere here?

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+
+                        @Override
+                        public void onSuccess(Void aVoid, HttpHeaders responseHeaders) throws IOException {
+                            logger.trace("successfully deleted blob [{}] in bucket [{}]", blobName, bucketName);
+                        }
+                    });
+
+                    count += 1;
+                } catch (IOException e) {
+                    logger.error("Unable to queue delete request for blob [{}] in bucket [{}]", e, blobName, bucketName);
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java]: Maybe it warrants bailing out right here with an exception. That sounds pretty serious to me.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+                if ((count >= MAX_BATCHING_REQUESTS)  || (it.hasNext() == false)) {
+                    try {
+                        logger.trace("executing batch request with [{}] actions", batch.size());
+                        batch.execute();
+                        if (errors.size() > 0) {
+                            throw new IOException("Failed to delete [" + errors.size() + "] blobs out of [" + blobNames.size() + "]");
+                        }
+                    } finally {
+                        batch = null;
+                        errors.clear();
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java]: Why do we need to clean errors here?

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+        @Override
+        public void moveBlob(String sourceBucketName, String sourceBlobName, String targetBucketName, String targetBlobName) throws IOException {
+            logger.debug("moving blob from [{}/{}] to [{}/{}]", sourceBucketName, sourceBlobName, targetBucketName, targetBlobName);
+            client.objects().copy(sourceBucketName, sourceBlobName, targetBucketName, targetBlobName, null).execute();
+            deleteBlob(sourceBucketName, sourceBlobName);
+        }
+    }
+
+    private String buildKey(String keyPath, String s) {
+        return keyPath + (s == null ? "" : s);
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/InternalGoogleCloudStorageService.java]: Same here.

**Comment by imotov** [comment]
Left some comments. Let's figure out how to add some tests to this thing.

**Comment by codefx9** [comment]
Hi, is there any updates on this issue?

**Comment by chrislovecnm** [comment]
:1

**Comment by dakrone** [comment]
ping @tlrx, I think this needs comments to be addressed and then another review?

**Comment by tlrx** [comment]
> ping @tlrx, I think this needs comments to be addressed and then another review?

Sure, thanks for the kindly reminder @dakrone . Let's add an update to this pull request and some explanation why it takes so long to get in.

###### A bit of history
I started to work on this plugin when elasticsearch was around version 1.4. At this time, the Google Client library was almost incompatible with the way Snapshot/Restore worked in ES. We try to find a workaround for some time but the solution was hardly maintainable and had poor performance. Around ES 1.5 or 1.6 the internal snapshot/restore changed a bit and at the same time the Google library improved a lot.  Great! But... Elasticsearch was on its way to 2.x and a lot of things changed and improved too: we moved the plugins from their dedicated repository back in elasticsearch repo, then move from maven to gradle, enabled the security manager, changed the way integration tests worked for plugins. It has been a lot of amazing improvements but keeping this pull request up-to-date required a LOT of attention and time. Knowing that things would stabilize, I put this on a shelve, I strongly apologize for that.

###### Current status
Anyway, I rebased and updated the code today to make it work with elasticsearch 5.0.0-alpha1-SNAPSHOT. It has been tested locally and on Google Compute Engine and it works nicely.

If you want to test it, you can check out this pull request, build it locally using Gradle and deploy it in a elasticsearch 5.0.0-alpha1-SNAPSHOT (download it [here](https://oss.sonatype.org/content/repositories/snapshots/org/elasticsearch/distribution/zip/elasticsearch/5.0.0-alpha1-SNAPSHOT/)). This pull request contains a documentation page.

@imotov Can you please have another look when you have time? That would be great!

File: plugins/repository-gcs/licenses/google-NOTICE.txt [hunk 12]
```diff
@@ -0,0 +1 @@
+
```
Comment (imotov) [inline plugins/repository-gcs/licenses/google-NOTICE.txt]: Is this file intentionally empty?

File: plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+        this.keyPath = keyPath;
+    }
+
+    @Override
+    public boolean blobExists(String blobName) {
+        try {
+            return client.blobExists(bucket, buildKey(blobName));
+        } catch (GoogleJsonResponseException e) {
+            GoogleJsonError error = e.getDetails();
+            if ((e.getStatusCode() == HTTP_NOT_FOUND) || ((error != null) && (error.getCode() == HTTP_NOT_FOUND))) {
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java]: There are very similar checks in side `client.blobExists(...)`. Maybe just move all checks there?

File: plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        return client.listBlobsByPrefix(bucket, keyPath, prefix);
+    }
+
+    @Override
+    public InputStream readBlob(String blobName) throws IOException {
+        try {
+            return client.readBlob(bucket, buildKey(blobName));
+        } catch (GoogleJsonResponseException e) {
+            GoogleJsonError error = e.getDetails();
+            if ((e.getStatusCode() == HTTP_NOT_FOUND) || ((error != null) && (error.getCode() == HTTP_NOT_FOUND))) {
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java]: Or maybe move all checks here. It's just having some checks inside client and other inside container is confusion. I have slight preference for moving all checks like this inside the client. On the second thought why do we need client to start with? If we move all checks inside client this becomes very thin facade for the client, which basically passes everything into the client. So, in order to remove complexity, maybe it makes sense to get rid of client (which is also quite thin facade for Google API) and just move all code from client here.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+    GoogleCloudStorageBlobContainer(BlobPath path, String bucket, GoogleCloudStorageClient client) {
+        super(path);
+        this.client = client;
+        this.bucket = bucket;
+
+        String keyPath = path.buildAsString("/");
+        if (!keyPath.isEmpty()) {
+            keyPath = keyPath + "/";
+        }
+        this.keyPath = keyPath;
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java]: I have seen this code so many times. Maybe it's time to move it somewhere where it can be reused?

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClient.java [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+    public boolean doesBucketExist(String bucketName) throws IOException {
+        return doPrivileged(() -> {
+            try {
+                Bucket bucket = client.buckets().get(bucketName).execute();
+                if (bucket != null) {
+                    return Strings.hasText(bucket.getId());
+                }
+            } catch (GoogleJsonResponseException e) {
+                if (e.getStatusCode() == HTTP_FORBIDDEN) {
+                    throw new IOException("Unable to get bucket [" + bucketName + "] because of insufficient privileges");
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClient.java]: I am wondering if it's safe to pretend that bucket doesn't exist when we don't have sufficient privileges. I think it might be safer to just bailout here.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClient.java [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+    public boolean blobExists(String bucketName, String blobName) throws IOException {
+        return doPrivileged(() -> {
+            try {
+                StorageObject blob = client.objects().get(bucketName, blobName).execute();
+                if (blob != null) {
+                    return Strings.hasText(blob.getId());
+                }
+            } catch (GoogleJsonResponseException e) {
+                if (e.getStatusCode() == HTTP_FORBIDDEN) {
+                    return false;
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClient.java]: Same here. Just let this exception to be thrown.

**Comment by imotov** [comment]
Left a few comments. Looks pretty good. I think we can simplify thing by getting rid of `GoogleCloudStorageClient` and just move all its functionality directly into `GoogleCloudStorageBlobStore` and `GoogleCloudStorageBlobContainer`. Also, what about `ESBlobStoreRepositoryIntegTestCase`?

**Comment by haizaar** [comment]
> Current status
> 
> Anyway, I rebased and updated the code today to make it work with elasticsearch 5.0.0-alpha1-SNAPSHOT. It has been tested locally and on Google Compute Engine and it works nicely.
> 
> If you want to test it, you can check out this pull request, build it locally using Gradle and deploy it in a elasticsearch 5.0.0-alpha1-SNAPSHOT (download it here). This pull request contains a documentation page.

@tlrx thanks for the update. Do I read you right that the current plan is to release it as part of ES 5.0?

File: plugins/repository-gcs/licenses/google-NOTICE.txt (hunk 12 above)
Comment (tlrx) [inline plugins/repository-gcs/licenses/google-NOTICE.txt]: It looks like we add empty file for libs that don't provide NOTICE.txt (I spotted few others like this in other plugins)

File: plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java (hunk 14 above)
Comment (tlrx) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java]: I agree. I removed the `GoogleCloudStorageClient` and moved most of the implementation in the blob store. It's easier to read and to maintain.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java (hunk 15 above)
Comment (tlrx) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java]: OK. Maybe we could move it to the `BlobPath.buildAsString()` because it is always used with this logic. If we do that, it makes sense to do it in another PR because other plugin like `repository-s3` can benefit from it.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClient.java (hunk 16 above)
Comment (tlrx) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageClient.java]: Totally agree, I can't remember why I did this...

**Comment by tlrx** [comment]
@imotov Thanks for your review! I removed the `GoogleCloudStorageClient` class and moved the implementation in the blob store. 

Concerning the tests, I moved the `ESBlobStoreRepositoryIntegTestCase`, `ESBlobStoreContainerTestCase` and `ESBlobStoreTestCase` classes into the test framework so that they can be inherited in plugins (it looks like we already have such abstract test classes like `AbstractNumericTestCase` in the test framework). Then I created a dedicated test for each of theses classes. I had to turn the `GoogleCloudStorageService` into an interface so that it can be mocked and I implemented a `MockHttpTransport` that simulates the Google Cloud Storage service similarly to what is done for the `discovery-gce` plugin. I think we're good on testing side.

Can you please have another look? That would be great.

@haizaar I can't give you any estimated date for this but I'd be happy to see it ready and testable for a beta release of 5.

**Comment by haizaar** [comment]
Tanguy, thanks for the update. Really looking forward to it.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+        private final Environment environment;
+
+        @Inject
+        public InternalGoogleCloudStorageService(Settings settings, Environment environment) {
+            super(settings);
+            this.environment = environment;
+        }
+
+        @Override
+        public Storage client(String serviceAccount, String application, TimeValue connectTimeout, TimeValue readTimeout)
```
Comment (imotov) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java]: Maybe rename it to createClient?

**Comment by imotov** [comment]
Left a minor comment. LGTM.

File: plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java (hunk 18 above)
Comment (tlrx) [inline plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageService.java]: Sure

**Comment by tlrx** [comment]
Thanks @imotov !
